File: src/PHP/StructuralPatterns/Composite/WindowsFormsApplication/Leaves/BaseControlLeaf.php
<?php

namespace StructuralPatterns\Composite\WindowsFormsApplication\Leaves;

use StructuralPatterns\Composite\WindowsFormsApplication\Component\Control;

abstract class BaseControlLeaf extends Control
{
    public function __construct($name, $width, $height, $foreColor, $backColor)
    {
        parent::__construct($name, $width, $height, $foreColor, $backColor);
    }

    public function getControls(): array
    {
        return [];
    }

    public function add(Control $control): void
    {
        echo "Cannot add control to " . get_class($this) . PHP_EOL;
    }

    public function remove(Control $control): void
    {
        echo "Cannot remove control from " . get_class($this) . PHP_EOL;
    }

    public function getHierarchicalLevel($depthIndicator): string
    {
        return parent::getHierarchicalLevel($depthIndicator) . PHP_EOL;
    }
}

File: src/C#/StructuralPatterns/Composite/WindowsFormsApplication/Leaves/Abstraction/BaseControlLeaf.cs
using System;
using System.Collections.Generic;
using System.Drawing;
using WindowsFormsApplication.Component;

namespace WindowsFormsApplication.Leaves.Abstraction
{
    public abstract class BaseControlLeaf : Control
    {
        public BaseControlLeaf(string name, int width, int height, Color foreColor, Color backColor)
            : base(name, width, height, foreColor, backColor)
        {

        }

        public override ICollection<Control> Controls { get; set; } = null;

        public override void Add(Control control)
        {
            Console.WriteLine($"Cannot add control to {GetType().Name}");
        }

        public override void Remove(Control control)
        {
            Console.WriteLine($"Cannot remove control from {GetType().Name}");
        }

        public override string GetHierarchicalLevel(int depthIndicator)
        {
           return $"{new string('-', depthIndicator)}+ " +
                $"Name: {Name}, Width: {Width}, Fore Color: ({ForeColor.A}, {ForeColor.G}, {ForeColor.B}), " +
                $"Back Color: ({BackColor.A}, {BackColor.G}, {BackColor.B})\r\n";
        }
    }
}
